sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala

```diff
@@ -589,18 +590,34 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
         col.getType.startsWith(serdeConstants.CHAR_TYPE_NAME))
       .map(col => col.getName).toSet
 
-    filters.collect {
-      case op @ BinaryComparison(a: Attribute, Literal(v, _: IntegralType)) =>
-        s"${a.name} ${op.symbol} $v"
-      case op @ BinaryComparison(Literal(v, _: IntegralType), a: Attribute) =>
-        s"$v ${op.symbol} ${a.name}"
-      case op @ BinaryComparison(a: Attribute, Literal(v, _: StringType))
-          if !varcharKeys.contains(a.name) =>
-        s"""${a.name} ${op.symbol} ${quoteStringLiteral(v.toString)}"""
-      case op @ BinaryComparison(Literal(v, _: StringType), a: Attribute)
-          if !varcharKeys.contains(a.name) =>
-        s"""${quoteStringLiteral(v.toString)} ${op.symbol} ${a.name}"""
-    }.mkString(" and ")
+    def isFoldable(expr: Expression): Boolean =
+      (expr.dataType.isInstanceOf[IntegralType] || expr.dataType.isInstanceOf[StringType]) &&
+      expr.foldable &&
+      expr.deterministic
+
+    def convertFoldable(expr: Expression): String = expr.dataType match {
+      case _: IntegralType => expr.eval(null).toString
+      case _: StringType => quoteStringLiteral(expr.eval(null).toString)
+    }
+
+    def convert(filter: Expression): String =
+      filter match {
+        case In(a: Attribute, exprs) if exprs.forall(isFoldable) =>
+          val or = exprs.map(expr => s"${a.name} = ${convertFoldable(expr)}").reduce(_ + " or " + _)
```

Comment:
In testing this in a modified form of `HiveClientSuite`, Hive is complaining that it can't parse the predicate. Specifically, the error message in the exception it's throwing is

```
Error parsing partition filter : line 1:3 no viable alternative at character '`'
```

The filter string it's trying to parse is `` `ds` = 20170101``.